### ISSUE TITLE
Round Trip state -> Vec<u8> -> Update -> apply_update loses type information

### ISSUE THREAD

**Issue by favilo** [y-crdt/y-crdt#270]
When attempting to copy the full state from one document with sub-docs to a fully empty document, I wrote the following code:

```rust
    #[test]
    fn round_trip_sub_docs() {
        let outer = Doc::with_client_id(123);
        let array = outer.get_or_insert_array("array");
        let map = outer.get_or_insert_map("map");
        let mut txn = outer.transact_mut();

        let inner = Doc::new();
        let inner_text = inner.get_or_insert_text("text");
        inner_text.push(&mut inner.transact_mut(), "Testing");
        array.push_back(&mut txn, inner);

        let text2 = TextPrelim::new("Testing again with non-doc");
        map.insert(&mut txn, "text", text2);

        let bytes = txn.encode_state_as_update_v1(&StateVector::default());
        drop(txn);

        let other = Doc::with_client_id(321);
        other
            .transact_mut()
            .apply_update(Update::decode_v1(&bytes).unwrap());
        assert_eq!(
            outer.transact().root_refs().collect::<Vec<_>>(),
            other.transact().root_refs().collect::<Vec<_>>(),
        );
}
```

However, instead of getting equivalent documents out, the `root_refs()` iterator on the `other` document panics with `Cannot convert to value - unsupported type ref: 15`. And printing out the Debug information for both of those, shows that both `array` and `map` on the `<root>` object in the `other` doc lost type information.

I've been searching the code for what causes it, and from what I can tell, this is caused because we are using a Named TypePtr when applying the update [here](https://github.com/y-crdt/y-crdt/blob/31b309868006d14c47c5deac6875979245de6409/yrs/src/block.rs#L1488).

Of course I'm not familiar enough with the code to tell if that is actually where the problem lies. But I'd love to help fix it, if I could get some pointers on where the best place to look would be.

**Comment by Horusiath** [comment]
Yjs doesn't include information about root types (other than their name) as part of the payload - Yjs/Yrs types are interchangeable so you can ie. cast `XmlElementRef` into `MapRef` etc. as they work as a projections over the underlying branch type.

**tl;dr** version is that user should define all root types before using them, ideally right after the document creation. We're thinking about pushing that as a requirement into the API (so that user has to define all root types at the moment when calling `Doc::new`).

**Comment by favilo** [comment]
Ah, right, I'd missed that line in the docs. I would really appreciate that change to the API, too. I'll close this out. Thanks for the response, and I love the work you folks are doing on this library. It's very exciting stuff.
